flashlight/fl/nn/modules/Module.h

```diff
@@ -52,8 +52,11 @@ class Module {
    * Constructs a module given its parameters.
    *
    * @param params a vector of `Variable` which will replace `params_`
+   * @param train Sets the module to training mode if true, eval mode if false.
+   * This changes all parameters so that gradient calculation will be
+   * enabled/disabled for any calls to `forward`.
    */
-  explicit Module(const std::vector<Variable>& params);
+  explicit Module(const std::vector<Variable>& params, bool train = true);
```

Comment:
This was necessary to correctly handle custom copy constructors of derived modules. If a `Module` has train set to false and is then copied, the copied module would have train set to true, but the copied parameters would have their `sharedGrad_->calcGrad` set to false.

I did consider just setting the params to the default train = true in the constructor, but this would break cloning containers where some modules have gradients disabled. For example a pre-trained feature extractor module/container as part of a larger network.

Reply:
I feel the better way to do this would be to call `module.eval()` on the newly-constructed module if the original instance that is being copied is in train mode. Should be simple:
```cpp
if (other.isTrain()) {
  inst.train();
}
// symmetrically for eval
```

I'm also now realizing there's no way to _read_ whether a module is in train mode or not. Would you be willing to send a separate PR adding that and a few small tests?

Reply:
My concern is that if a user does not call `train()` or `eval()` after clone/copy for whatever reason, it allows the module to be in an invalid state such that the modules parameters `sharedGrad_->calcGrad` and `train_` can be inconsistent after cloning. For example:

```cpp
Linear lin(10,20);
lin.eval();

auto linCopy = lin; // linCopy params gradients are disabled, but train_ = true
// now either train() or eval() must be called depending on the desired behaviour
```
Hence *every* time a copy is made (via clone or via C++ value semantics like above) the user must ensure the correct train mode is set. If it is not done, it could lead to subtle bugs where gradients aren't correctly propagated and its hard to identify the cause, or gradients are collected at inference unnecessarily.

I think it also just makes sense that a cloned/copied module should be in an identical state as its original and shouldn't require a user to take action to ensure it is. I think it's also a good idea to prevent objects from entering an invalid state under normal use.

Reply:
Perhaps an alternative approach might resolve this? It would essentially involve checking the parameters and setting `train_ = true` if any parameters have gradients enabled, so something like this:

```cpp
Module::Module(const std::vector<Variable>& params)
    : params_(params.begin(), params.end()),
        train_(std::any_of(params.begin(), params.end(),
            [](const Variable& param) { return param.isCalcGrad(); })
        ) {}
```

This has the advantage of ensuring the module is not in an invalid state when constructed, avoids adding `train` as a constructor parameter and does not call `eval()` or `train()` in the constructor. Users can then call `eval()` or `train()` based on their use case.

Thoughts on this approach?

Reply:
Following on from the other comment chain regarding `cloneModule`, I think maybe we've gotten lost in the weeds. Given that the constructor `Module::Module(const std::vector<Variable>& params)` is protected and only derived modules can call it, the derived module can also just set `train_` accordingly in its constructor. For example see `fl/nn/modules/Linear.cpp`:

```cpp
Linear::Linear(const Linear& other)
    : UnaryModule(other.copyParams()),
      nIn_(other.nIn_),
      nOut_(other.nOut_),
      bias_(other.bias_) {
  train_ = other.train_; // <----- train is set here
}
Linear& Linear::operator=(const Linear& other) {
  params_ = other.copyParams();
  train_ = other.train_; // <----- train is set here
  nIn_ = other.nIn_;
  nOut_ = other.nOut_;
  bias_ = other.bias_;
  return *this;
}
```

This can also be done in the copy assignment operator for the derived Module. I've originally updated most Modules in flashlight to do this in the PR. While reviewing the PR changes, I found a few places where I missed it, hence my intention with setting `train_` in the constructor of `Module` was to try remove this source of bug where users forget to set it in their derived Module copy constructor. However, I've just realised this doesn't address the copy assignment operator case. So given users need to copy `params_` in their copy constructors anyway, I believe it would be better to be more consistent and simply require users also set `train_` in both their copy constructors and copy assignment operators and leave the `Module` constructor as it was originally.